structuretoolkit/analyse/snap.py

```diff
@@ -0,0 +1,397 @@
+from ctypes import c_double, c_int, cast, POINTER
+import numpy as np
+from scipy.constants import physical_constants
+
+eV_div_A3_to_bar = 1e25 / physical_constants["joule-electron volt relationship"][0]
+
+
+def convert_mat(mat):
+    mat[np.diag_indices_from(mat)] /= 2
+    return mat[np.triu_indices(len(mat))]
+
+
+def calc_per_atom_quad(linear_per_atom):
+    return np.array(
+        [
+            np.concatenate(
+                (
+                    atom,
+                    convert_mat(
+                        mat=np.dot(
+                            atom.reshape(len(atom), 1), atom.reshape(len(atom), 1).T
+                        )
+                    ),
+                )
+            )
+            for atom in linear_per_atom
+        ]
+    )
+
+
+def calc_sum_quad(linear_sum):
+    return np.concatenate(
+        (
+            linear_sum,
+            convert_mat(
+                mat=np.dot(
+                    linear_sum.reshape(len(linear_sum), 1),
+                    linear_sum.reshape(len(linear_sum), 1).T,
+                )
+            ),
+        )
+    )
+
+
+def get_apre(cell):
+    a, b, c = cell
+    an, bn, cn = [np.linalg.norm(v) for v in cell]
+
+    alpha = np.arccos(np.dot(b, c) / (bn * cn))
+    beta = np.arccos(np.dot(a, c) / (an * cn))
+    gamma = np.arccos(np.dot(a, b) / (an * bn))
+
+    xhi = an
+    xyp = np.cos(gamma) * bn
+    yhi = np.sin(gamma) * bn
+    xzp = np.cos(beta) * cn
+    yzp = (bn * cn * np.cos(alpha) - xyp * xzp) / yhi
+    zhi = np.sqrt(cn**2 - xzp**2 - yzp**2)
+
+    return np.array(((xhi, 0, 0), (xyp, yhi, 0), (xzp, yzp, zhi)))
+
+
+def write_ase_structure(lmp, structure):
+    number_species = len(set(structure.get_chemical_symbols()))
+
+    apre = get_apre(cell=structure.cell)
+    (
+        (xhi, xy, xz),
+        (_, yhi, yz),
+        (_, _, zhi),
+    ) = apre.T
+    lmp.command(
+        "region 1 prism"
+        + " 0.0 "
+        + str(xhi)
+        + " 0.0 "
+        + str(yhi)
+        + " 0.0 "
+        + str(zhi)
+        + " "
+        + str(xy)
+        + " "
+        + str(xz)
+        + " "
+        + str(yz)
+        + " units box"
+    )
+    lmp.command("create_box " + str(number_species) + " 1")
+
+    el_dict = {el: i for i, el in enumerate(set(structure.get_chemical_symbols()))}
+
+    R = np.dot(np.linalg.inv(structure.cell), apre)
+
+    positions = structure.positions.flatten()
+    if np.matrix.trace(R) != 3:
+        positions = np.array(positions).reshape(-1, 3)
+        positions = np.matmul(positions, R)
+    positions = positions.flatten()
```

Comment:
Maybe I'm not following the array shape closely enough, but it feels that some of the flattens are redundant.